src/imitation/scripts/train_adversarial.py

```diff
@@ -236,20 +177,136 @@ def should_plot_now(epoch) -> bool:
+      reward_type = "DiscrimNetGAIL"
+    else:
+      raise RuntimeError(f"Unknown reward type for {trainer.discrim}")
+
+    return dict(ep_reward_mean=ep_reward_mean,
+                ep_reward_std_err=ep_reward_std_err,
+                log_dir=log_dir,
+                transfer_reward_path=reward_path,
+                transfer_reward_type=reward_type)
+
+
+class _TrainVisualizer:
+  def __init__(self,
+               trainer: "imitation.algorithms.adversarial.AdversarialTrainer",
+               show_plots: bool,
+               n_episodes_per_reward_data: int,
+               log_dir: str,
+               expert_policy=None):
+    """
+    Args:
+      trainer: AdversarialTrainer used to perform rollouts.
+      show_plots: If True, then `plt.show()` plot in addition to saving them.
+      n_episodes_per_reward_data: The number of episodes to rollout out when
+        determining the mean episode reward.
+      n_epochs_per_plot: An optional number, greater than or equal to 1. The
+        (possibly fractional) number of epochs between each plot. The first
+        plot is at epoch 0, after the first discrim and generator steps.
+        If `n_epochs_per_plot is None`, then don't make any plots.
+      expert_policy (BasePolicy or BaseRLModel, optional): If provided,
+          then also plot the performance of this expert policy.
+    """
+    self.trainer = trainer
+    self.show_plots = show_plots
+    self.n_episodes_per_reward_data = n_episodes_per_reward_data
+    self.log_dir = log_dir
+    self.expert_policy = expert_policy
+    self.plot_idx = 0
+    self.gen_data = ([], [])
+    self.disc_data = ([], [])
+
+    self.gen_ep_reward = defaultdict(list)
+    self.rand_ep_reward = defaultdict(list)
+    self.exp_ep_reward = defaultdict(list)
 
-def _savefig_timestamp(prefix="", also_show=True):
-  path = "output/{}_{}.png".format(prefix, datetime.datetime.now())
-  plt.savefig(path)
-  tf.logging.info("plot saved to {}".format(path))
-  if also_show:
-    plt.show()
+    # Collect data for epoch 0.
+    self.disc_plot_add_data(False)
+    self.ep_reward_plot_add_data(self.trainer.env, "Ground Truth Reward")
+    self.ep_reward_plot_add_data(self.trainer.env_train, "Train Reward")
+    self.ep_reward_plot_add_data(self.trainer.env_test, "Test Reward")
+
+  def disc_plot_add_data(self, gen_mode: bool = False):
+    """Evaluates and records the discriminator loss for plotting later.
+
+    Args:
+        gen_mode: Whether the generator or the discriminator is active.
+            We use this to color the data points.
+    """
+    mode = "gen" if gen_mode else "dis"
+    X, Y = self.gen_data if gen_mode else self.disc_data
+    # Divide by two since we get two data points (gen and disc) per epoch.
+    X.append(self.plot_idx / 2)
+    Y.append(self.trainer.eval_disc_loss())
+    tf.logging.info(
+        "plot idx ({}): {} disc loss: {}"
+        .format(mode, self.plot_idx, Y[-1]))
+    self.plot_idx += 1
+
+  def disc_plot_show(self):
```

Comment:
Method name is a bit misleading when it may never call `plt.show()`. Maybe just `plot_disc`?

Reply:
I renamed to `plot_disc_loss`. Also renamed some other stuff in the class for consistency/clarity.